tests/system/test_misc.c

```diff
@@ -389,3 +389,88 @@ int cmd_exit_cb(int argc, char **argv)
    DEVSHELL_CMD_ASSERT(after_cb == before_cb + 1);
    return 0;
 }
+
+static bool
+comes_after(struct timeval t2,
+            struct timeval t1)
+{
+   // t1 must come before t2
+   return t2.tv_sec > t1.tv_sec || (t2.tv_sec == t1.tv_sec && t2.tv_usec > t1.tv_usec);
+}
+
+#define BEGIN_OF_TIME (struct timeval) {0, 0}
+
+static void busy_wait(void)
+{
+   // It's assumed that this loop takes
+   // more than 1us of cpu time.
+   for (int i = 0; i < 1000000; i++)
+      asm(""); // Don't let the compiler optimize this loop
+}
+
+static void busy_wait_kernel(void)
+{
+   // It's assumed that this loop takes
+   // more than 1us of cpu time in kernel
+   // space.
+   for (int i = 0; i < 100; i++) {
+      int n = fork();
```

Comment:
I appreciate the attempt, but this is the incorrect way of doing that.
1. Note that that the timer frequency (50-250 HZ), not the 1us precision of the timespec struct.
2. You're not waiting for all the child to exit before exiting from here. That very likely causes all the tests to fail: https://vkvaltchev.visualstudio.com/Tilck/_build/results?buildId=1959&view=logs&jobId=28f1ef74-9f3f-5899-b019-7358c934de06&j=b80d80ca-e3a4-5429-e1b2-036d7cb8925c&t=b6270386-105a-5b35-2e48-584241c3a271
3. fork() is not the right tool. Use a custom Tilck syscall wait kernel time. See TILCK_CMD_SYSCALL, sys_tilck_cmd and the kernel/tilck_cmd.c file. There, you can add an empty slot for a TILCK test syscall, and then set the right function pointer in modules/systests/systests.c with the implementation. 4. In the test TILCK syscall, just burn some cycles with a loop and NOP. But don't do it too much, we don't need to make tests unnecessarily slower.

Reply:
> Note that that the timer frequency (50-250 HZ), not the 1us precision of the timespec struct.

I don't think I understand. Are you saying I should consider the resolution of the timer instead of the `struct timeval`?

Reply:
`struct timeval` is the structure of the interface, but that doesn't mean that the operating system guarantees a resolution is small as what can be represented in that struct. Of course, not even Linux does. It depends on the HW and how the time spent is calculated. On newer machines (with "newer" I typically mean machines after 2004/2005, maybe 2008 not sure), it is possible to estimate more precisely the CPU time we spent running a task on a CPU thanks to the HPET. But Tilck still uses the most traditional and primitive way of calculating the time spent in tasks, which is based on the scheduler "ticks". Each tick happens with the same frequency, determined by the primary timer. Again, Linux now supports variable ticks, tickless mode etc. But, traditionally, that's the simplest thing it can be done. Therefore, with a freq of 250 HZ the smallest time quantum we can appreciate is 4 ms. Consider that's relatively good. On old machines in the early 90s, the freq was 100 HZ and that time was 10ms. 

Therefore, when writing loops like that, consider that you wanna burn at few quanta. Not less because otherwise there could be less than 1 tick, but not more either, not not waste resources. Also, in the test itself, consider that in Azure sometimes VMs get stuck or can run very slowly. Therefore, it's possible that a long loop can take much more time than on a regular machine, in some corner cases.